object/slicer/slicer.go

```diff
@@ -0,0 +1,255 @@
+package slicer
+
+import (
+	"crypto/sha256"
+	"fmt"
+	"hash"
+	"io"
+
+	"github.com/nspcc-dev/neofs-sdk-go/checksum"
+	cid "github.com/nspcc-dev/neofs-sdk-go/container/id"
+	"github.com/nspcc-dev/neofs-sdk-go/object"
+	oid "github.com/nspcc-dev/neofs-sdk-go/object/id"
+	"github.com/nspcc-dev/neofs-sdk-go/user"
+	"github.com/nspcc-dev/neofs-sdk-go/version"
+	"github.com/nspcc-dev/tzhash/tz"
+)
+
+// ObjectWriter represents a virtual object recorder.
+type ObjectWriter interface {
+	// InitDataStream initializes and returns a stream of writable data associated
+	// with the object according to its header. Provided header includes at least
+	// owner, container and object ID fields.
+	InitDataStream(header object.Object) (dataStream io.Writer, err error)
+}
+
+// SliceDataIntoObjects creates new NeoFS object from the input data stream,
+// associates the object with the given container, makes it owned by the
+// specified user, and writes the object to the given ObjectWriter. After a
+// successful write, SliceDataIntoObjects returns an oid.ID which is a unique
+// reference to the object in the container.
+//
+// SliceDataIntoObjects sets all required calculated fields like payload length,
+// checksum, etc.
+//
+// If ObjectWriter provides io.Closer, it is closed after the payload has been
+// written. In this case, SliceDataIntoObjects fails immediately on Close error.
+//
+// By default, single object with empy payload is created. Various Options allow
+// you to tune the behavior.
+func SliceDataIntoObjects(cnr cid.ID, owner user.ID, w ObjectWriter, opts Options) (rootID oid.ID, err error) {
+	// FIXME: can't copy object.Object instances
+	fillCommonMetadata := func(obj *object.Object) {
+		currentVersion := version.Current()
+		obj.SetVersion(&currentVersion)
+		obj.SetContainerID(cnr)
+		obj.SetOwnerID(&owner)
+		obj.SetCreationEpoch(opts.currentNeoFSEpoch)
+		obj.SetType(object.TypeRegular)
+		obj.SetSessionToken(opts.sessionToken)
+	}
+
+	var rootHeader object.Object
+	fillCommonMetadata(&rootHeader)
+
+	if opts.objectPayloadLimit == 0 {
+		opts.objectPayloadLimit = 1 << 20
+		// TODO: if it's larger than the actual limit, we handle server error and resize
```

Comment:
The problem is we can't fetch size limit from the server properly. We can either:
1. to accept some `type Settings interface` which encapsulates values like max size, current, epoch, etc.
2. support overflow status code in the protocol

IMO 2 is better, but 1 will also work.

Reply:
Don't we have NetworkInfo requests to get this data? I think if we're talking neofs-_sdk_ here we want to provide some simple interfaces that work out of the box. If some aspects of this function require getting network info then that's what `slicer` should do behind the scenes. Give an ability to override this (for all three people that care or have this data already for some reason and wanting to avoid useless requests), but make it work by default, I think that's the best combo.

Take NeoGo RPC client as an example, it needs to get some server settings (like stateroot), so that's what it does in `Init`, no one cares much about what's happening there, people just call it and then expect everything else to work (as it does).

Reply:
Overall, I agree with you. But this package is intended to be directly used only by those three people who implement server side, and the `client` package. Normally clients won't use `slicer` package at all since this functionality will lie under `client.Put`. My own goal is to not teach SDK users how to slice the objects but let them use client as they did before.